sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Product.scala

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions.aggregate
+
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.dsl.expressions._
+import org.apache.spark.sql.catalyst.expressions.{AttributeReference, Expression, ImplicitCastInputTypes, Literal}
+import org.apache.spark.sql.catalyst.util.TypeUtils
+import org.apache.spark.sql.types.{AbstractDataType, DataType, DoubleType}
+
+
+/** Multiply numerical values within an aggregation group */
+case class Product(child: Expression)
+    extends DeclarativeAggregate with ImplicitCastInputTypes {
+
+  override def children: Seq[Expression] = child :: Nil
+
+  override def nullable: Boolean = true
+
+  override def dataType: DataType = resultType
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(DoubleType)
+
+  override def checkInputDataTypes(): TypeCheckResult =
+    TypeUtils.checkForNumericExpr(child.dataType, "function product")
+
+  private val resultType = DoubleType
+
+  private lazy val product = AttributeReference("product", resultType)()
+
+  private lazy val one = Literal(1.0, resultType)
+
+  override lazy val aggBufferAttributes = product :: Nil
+
+  override lazy val initialValues: Seq[Expression] =
+    Seq(Literal(null, resultType))
+
+  override lazy val updateExpressions: Seq[Expression] = {
+    // Treat the result as null until we have seen at least one child value,
+    // whereupon the previous product is promoted to being unity.
+    // Each child value is implicitly multiplied by a scaling factor
+    // (1.0 by default) so that the client has some control over overflow
```

Comment:
sorry I may missed something. Why do we need to multiple by 1.0? double * 1.0 means noop IIUC.

Reply:
Actually this doesn't match the code. We simply multiple the child and the current product value.

Reply:
Can we update the comment here?